Add documentation of some table column APIs in Storybook

jattasNI opened the pull request:
# Pull Request

## 🤨 Rationale

We've been adding new APIs to table columns in general and to specific column types like text but haven't been documenting them. This handles some of the tasks in #1044.

## 👩‍💻 Implementation

- add a new story dedicated to table columns. new story includes docs for:
    - column order
    - column header content
    - column sorting
    - other common attributes that don't merit their own story: field name and placeholder
- add a new story for the text column type
- remove some content from the existing table story
- Also did some minor refactoring to increase code sharing

There are additional stories tracked in #1044 that I'm not tackling yet like docs for action menu and column sizing.

## 🧪 Testing

Visual inspection of new stories

## ✅ Checklist

<!--- Review the list and put an x in the boxes that apply or ~~strike through~~ around items that don't (along with an explanation). -->

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

rajsite commented on packages/nimble-components/src/table-column/tests/table-column.stories.ts:
```diff
@@ -0,0 +1,416 @@
+See the **Table** story for information about configuring the table itself.`;
+
+const metadata: Meta<CommonTableArgs> = {
+    title: 'Table Columns',
```
What do you think about having this render under Table in the navigation tree? Kinda how Menu Item is under menu

I kinda like the organization of `Table/Column Configuration`:
![image](https://user-images.githubusercontent.com/1588923/224375102-4c1e4523-c8f8-4027-9ed1-5f8a1ef32b48.png)

jattasNI replied:
I like the idea but I can't figure out how to get it to work. If I use the `Table/Column Configuration` naming as described in [Storybook docs](https://storybook.js.org/docs/react/writing-stories/naming-components-and-hierarchy#grouping) I end up with a separate top-level section (sibling to Tokens and Tests). I did briefly see the purple folder layout in your screenshot but that only seems to appear after a hot reload on localhost; if I refresh the page after that I get the separate section.

<img width="215" alt="image" src="https://user-images.githubusercontent.com/10500124/224448435-8ad6db51-b3fc-44b3-b83b-bbaaa05ad0fa.png">

I think a separate section for the table is worse than having multiple pages for column configuration so would vote to keep things as-is if we can't figure out how to get the layout in your screenshot to work reliably.

rajsite replied:
I'm actually kinda feeling the table as a separate section. It's a sufficiently complex component with potentially dozen of interrelated elements to be aware of and lots of nuance.
But it's also a comparatively minor organization point and not worth blocking this PR. I'd rather the PR go in sooner so we can start capturing docs as features are merged and refine the organization more later.

rajsite commented on packages/nimble-components/src/table-column/tests/table-column.stories.ts:
```diff
@@ -0,0 +1,416 @@
+import { html, ref, when } from '@microsoft/fast-element';
```
What about moving these files to table-column\base\tests? I think that's what these stories represent, the base features shared across all columns.

rajsite commented on packages/nimble-components/src/table-column/tests/table-column.stories.ts:
```diff
@@ -0,0 +1,416 @@
+        }
+    },
+    args: {
+        tableRef: undefined,
```
Can you add some entry to the args table that shows something. Even a field that's like "Minimal Column Configuration Example" or something.

Right now from the Canvas view if I click it, I don't know what I'm supposed to take away from it:
![image](https://user-images.githubusercontent.com/1588923/224373684-6b59075f-9700-4264-b1cf-ce6052b93ef4.png)

jattasNI replied:
I went back and forth on removing this story entirely since none of our other components that expose multiple stories do anything to address this. In the end I went with your suggestion but hope we can move away from the autogenerated docs tab and towards custom mdx docs someday. Then presumably we'd remove this placeholder story since I agree it is confusing on the Canvas tab.

rajsite commented on packages/nimble-components/src/table-column/tests/table-column.stories.ts:
```diff
@@ -0,0 +1,416 @@
+    },
+    // prettier-ignore
+    render: createUserSelectedThemeStory(html<ColumnOrderTableArgs>`
+        ${usageWarning('table')}
```
I think we should avoid having the usage warning render in the example code text. I'd argue the banner is enough in the rendered component and this just muddles trying to understand how it works. Maybe we can add the code-hide class to it?
![image](https://user-images.githubusercontent.com/1588923/224375543-0bc92895-1ed2-4549-8942-a62b41d975de.png)

jattasNI commented:
@atmgrifter00 @rajsite while addressing review feedback I ended up adding a couple more stories: 
1. Table Column Configuration/Common Attributes to cover `column-hidden` and `column-id`
2. A story for `nimble-table-column-text` APIs `field-name` and `placeholder-text`

I'm resetting Meyer to get another set of eyes on the doc content. Milan is welcome to review too but didn't reset you (hoping the changes are uncontroversial so I can get this in quicker and save you review time).

rajsite commented on packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+};
+
+export const commonTableArgs = (
```
Can you follow a pattern similar to the `sharedMatrixParameters` pattern.
In-specific:
- have the resulting object be `as const`
- Don't use the shared value directly but "merge" it. ie the Args type should extend the CommonTableArgs type, spread the common table args into args, spread the common tableargstypes into argtypes, etc. This makes it easier for future story authors to see where to insert new args, etc.
- Be consistent with the naming convention, i.e. the shared prefix.

rajsite commented on packages/nimble-components/src/table-column/base/tests/table-column-stories-utils.ts:
```diff
@@ -0,0 +1,36 @@
+            disable: true
+        }
+    }
+};
```
`as const` and spread into usage spots

rajsite commented:
> @atmgrifter00 @rajsite while addressing review feedback I ended up adding a couple more stories:
> 
> 1. Table Column Configuration/Common Attributes to cover `column-hidden` and `column-id`
> 2. A story for `nimble-table-column-text` APIs `field-name` and `placeholder-text`
> 
> I'm resetting Meyer to get another set of eyes on the doc content. Milan is welcome to review too but didn't reset you (hoping the changes are uncontroversial so I can get this in quicker and save you review time).

Just reviewed the code sharing pattern to match with existing patterns. Leaving review the actual doc content for the added stories to y'all.

atmgrifter00 commented on packages/nimble-components/src/table-column/base/tests/table-column.stories.ts:
```diff
@@ -0,0 +1,454 @@
+const headerContentIntro = 'The content of each column header comes from whatever is slotted in the column element.';
+const headerTextContent = `If you provide only text content, Nimble will style it
+and add a \`title\` to show a tooltip when truncated.`;
+const headerIconContent = 'If you provide icon content, you should set your own `title`.';
```
```suggestion
const headerIconContent = 'If you provide icon content, you should set your own `title` on the icon element.';
```
